File: jetpack_java/src/main/java/com/kunminx/jetpack_java/sample_05_navigation/ui/NavigationDetailFragment.java


package com.kunminx.jetpack_java.sample_05_navigation.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.kunminx.jetpack_java.R;
import com.kunminx.jetpack_java.common_data.Configs;
import com.kunminx.jetpack_java.common_data.bean.Moment;
import com.kunminx.jetpack_java.databinding.FragmentDetailNavigationBinding;
import com.kunminx.jetpack_java.sample_04_databinding.ui.state.DetailViewModel;
import com.kunminx.architecture.ui.BaseFragment;

/**
 * TODO：本示例专注于提供 Navigation 组件的使用场景和示例，因而其他内容均保持 Android 引入 Jetpack 前的模样，
 * kotlin 模块提供同样的场景和基于 Kotlin 的写法，可对照查阅。
 * 并且本项目的示例从 sample_01 到 sample_05 是循序渐进地 Jetpack 化，
 * 如看完对 Jetpack 高频核心组件 "为什么要用"、"为什么要这样用" 有了一丝丝好奇心，可前往《Jetpack MVVM 精讲》和《Jetpack MVVM 最佳实践》项目查阅深度解析。
 * <p>
 * https://juejin.im/post/5dafc49b6fb9a04e17209922
 * <p>
 * https://github.com/KunMinX/Jetpack-MVVM-Best-Practice
 * <p>
 * <p>
 * Create by KunMinX at 2020/5/30
 */
public class NavigationDetailFragment extends BaseFragment {

    private DetailViewModel mDetailState;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mDetailState = getFragmentScopeViewModel(DetailViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_detail_navigation, container, false);
        FragmentDetailNavigationBinding binding = FragmentDetailNavigationBinding.bind(view);
        binding.setLifecycleOwner(this);
        binding.setVm(mDetailState);
        binding.setClick(new ClickProxy());
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Moment moment = (Moment) getArguments().getParcelable(Configs.THIS_MOMENT);
        mDetailState.initState(moment);
    }

    public class ClickProxy {
        public void back() {
            nav().navigateUp();
        }
    }
}

File: jetpack_kotlin/src/main/java/com/flywith24/jetpack_kotlin/sample_05_navigation/ui/NavigationDetailFragment.kt
package com.flywith24.jetpack_kotlin.sample_05_navigation.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import com.flywith24.jetpack_kotlin.R
import com.flywith24.jetpack_kotlin.common_data.Configs
import com.flywith24.jetpack_kotlin.common_data.bean.Moment
import com.flywith24.jetpack_kotlin.databinding.KotlinFragmentDetailNavigationBinding
import com.flywith24.jetpack_kotlin.sample_04_databinding.ui.state.DetailViewModel
import com.kunminx.architecture.ui.BaseFragment

/**
 * @author Flywith24
 * @date   2020/6/1
 * time   22:06
 * description
 */
class NavigationDetailFragment : BaseFragment() {
    /**
     * fragment-ktx 扩展函数
     */
    private val mDetailViewModel by viewModels<DetailViewModel>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.kotlin_fragment_detail_navigation, container, false)
        val binding = KotlinFragmentDetailNavigationBinding.bind(view)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.vm = mDetailViewModel
        binding.click = ClickProxy()
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val moment = arguments?.getParcelable<Moment>(Configs.THIS_MOMENT)
        moment?.let {
            mDetailViewModel.initState(it)
        }
    }

    inner class ClickProxy {
        fun back() {
            nav().navigateUp()
        }
    }
}